File: ios/Sources/CapacitorAmplitude/AmplitudeAnalyticsPlugin.swift
import Foundation
import Capacitor
import AmplitudeSwift

@objc(AmplitudeAnalytics)
public class AmplitudeAnalyticsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AmplitudeAnalytics"
    public let jsName = "AmplitudeAnalytics"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "init", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "track", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reset", returnType: CAPPluginReturnPromise),
    ]

    private var amplitude: Amplitude?

    @objc func `init`(_ call: CAPPluginCall) {
        guard let apiKey = call.getString("apiKey"), !apiKey.isEmpty else {
            call.reject("Missing apiKey")
            return
        }

        if amplitude != nil {
            call.resolve()
            return
        }

        let serverZoneString = call.getString("serverZone") ?? "US"
        let serverZone: ServerZone = serverZoneString == "EU" ? .EU : .US

        let logLevelString = call.getString("logLevel") ?? "ERROR"
        let logLevel: LogLevelEnum
        switch logLevelString {
        case "OFF": logLevel = .OFF
        case "WARN": logLevel = .WARN
        case "INFO": logLevel = .INFO
        case "DEBUG": logLevel = .DEBUG
        default: logLevel = .ERROR
        }

        let config = Configuration(
            apiKey: apiKey,
            logLevel: logLevel,
            serverZone: serverZone
        )
        amplitude = Amplitude(configuration: config)
        call.resolve()
    }

    @objc func setUserId(_ call: CAPPluginCall) {
        // userId can be explicitly null to clear
        let userId: String? = call.getString("userId")
        amplitude?.setUserId(userId: userId)
        call.resolve()
    }

    @objc func track(_ call: CAPPluginCall) {
        guard let eventType = call.getString("eventType"), !eventType.isEmpty else {
            call.reject("Missing eventType")
            return
        }

        let properties = call.getObject("eventProperties")
        let event = BaseEvent(eventType: eventType)
        if let props = properties {
            var eventProps: [String: Any] = [:]
            for (key, value) in props {
                eventProps[key] = value
            }
            event.eventProperties = eventProps
        }
        amplitude?.track(event: event)
        call.resolve()
    }

    @objc func reset(_ call: CAPPluginCall) {
        amplitude?.reset()
        call.resolve()
    }
}

File: android/src/main/java/de/bsdex/capacitor/amplitude/AmplitudeAnalyticsPlugin.java
package de.bsdex.capacitor.amplitude;

import com.amplitude.android.Amplitude;
import com.amplitude.android.Configuration;
import com.amplitude.core.ServerZone;
import com.amplitude.core.events.BaseEvent;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

@CapacitorPlugin(name = "AmplitudeAnalytics")
public class AmplitudeAnalyticsPlugin extends Plugin {

    private Amplitude amplitude;

    @PluginMethod()
    public void init(PluginCall call) {
        String apiKey = call.getString("apiKey");
        if (apiKey == null || apiKey.isEmpty()) {
            call.reject("Missing apiKey");
            return;
        }

        if (amplitude != null) {
            call.resolve();
            return;
        }

        String serverZoneStr = call.getString("serverZone", "US");
        ServerZone zone = "EU".equals(serverZoneStr) ? ServerZone.EU : ServerZone.US;

        Configuration config = new Configuration(apiKey, getContext().getApplicationContext());
        config.setServerZone(zone);

        amplitude = new Amplitude(config);
        call.resolve();
    }

    @PluginMethod()
    public void setUserId(PluginCall call) {
        String userId = call.getString("userId");
        if (amplitude != null) {
            amplitude.setUserId(userId);
        }
        call.resolve();
    }

    @PluginMethod()
    public void track(PluginCall call) {
        String eventType = call.getString("eventType");
        if (eventType == null || eventType.isEmpty()) {
            call.reject("Missing eventType");
            return;
        }

        if (amplitude == null) {
            call.reject("Amplitude not initialized. Call init() first.");
            return;
        }

        BaseEvent event = new BaseEvent();
        event.eventType = eventType;

        JSObject props = call.getObject("eventProperties");
        if (props != null) {
            Map<String, Object> propsMap = new HashMap<>();
            Iterator<String> keys = props.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                try {
                    propsMap.put(key, props.get(key));
                } catch (Exception e) {
                    // skip unparseable values
                }
            }
            event.setEventProperties(propsMap);
        }

        amplitude.track(event);
        call.resolve();
    }

    @PluginMethod()
    public void reset(PluginCall call) {
        if (amplitude != null) {
            amplitude.reset();
        }
        call.resolve();
    }
}
